instaclustr/resource_kafka_topic.go

```diff
@@ -0,0 +1,382 @@
+	if len(d.Get("config").([]interface{})) == 0 {
+		log.Printf("[INFO] There are no configs to be updated.")
+		return nil
+	}
+	client := meta.(*Config).Client
+
+	changedConfigMap, err := getChangedConfigMap(d)
+	if err != nil {
+		return err
+	}
+	if len(changedConfigMap) == 0 {
+		log.Printf("[INFO] There are no configs to be updated.")
+		err = resourceKafkaTopicRead(d, meta)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+
+	var kafkaTopicConfigOptions KafkaTopicConfigOptions
+	err = mapstructure.Decode(changedConfigMap, &kafkaTopicConfigOptions)
+	if err != nil {
+		return err
+	}
+
+	updateKafkaTopicRequest := UpdateKafkaTopicRequest{
+		Config: &kafkaTopicConfigOptions,
+	}
+
+	jsonStr, err := json.Marshal(updateKafkaTopicRequest)
+	if err != nil {
+		return fmt.Errorf("[Error] Error creating kafka topic update request: %s", err)
+	}
+
+	err = client.UpdateKafkaTopic(cluster_id, topic, jsonStr)
+	if err != nil {
+		return fmt.Errorf("[Error] Error updating the config of Kafka topic %s: %s", topic, err)
+	}
+
+	//Reading updated configs
+	err = resourceKafkaTopicRead(d, meta)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[INFO] The configs of Kafka topic %s has been updated.", topic)
+	return nil
+}
+
+func getChangedConfigMap(d *schema.ResourceData) (map[string]interface{}, error) {
+	newConfig := d.Get("config").([]interface{})[0].(map[string]interface{})
+	changedConfigMap := make(map[string]interface{}, 0)
+	var kafkaTopicConfig KafkaTopicConfigOptions
+	decodedConfigMap := make(map[string]interface{}, 0) // This map stores all non-zero values and all the pointers
+
+	// Decode twice here because zero-value non-pointers need to be removed
+	err := mapstructure.Decode(newConfig, &kafkaTopicConfig)
+	if err != nil {
+		return nil, err
+	}
+	err = mapstructure.Decode(kafkaTopicConfig, &decodedConfigMap)
+	if err != nil {
+		return nil, err
+	}
+
+	// The following block is to get what configs are indeed changed
+	for k, v := range newConfig {
+		key := fmt.Sprintf("config.0.%s", k)
+		valueOfV := reflect.ValueOf(decodedConfigMap[k])
+		if d.HasChange(key) {
+			before, after := d.GetChange(key)
+			isBeforeZeroValue := before == reflect.Zero(reflect.TypeOf(before)).Interface()
+			isAfterZeroValue := after == reflect.Zero(reflect.TypeOf(after)).Interface()
+			if !isBeforeZeroValue && !isAfterZeroValue {
+				changedConfigMap[k] = v
+			}
+			if !isBeforeZeroValue && isAfterZeroValue && (valueOfV.Kind() == reflect.Ptr) {
+				changedConfigMap[k] = v
+			}
+			if isBeforeZeroValue && !isAfterZeroValue {
+				changedConfigMap[k] = v
+			}
+		}
+	}
+	return changedConfigMap, nil
+}
+
+func resourceKafkaTopicDelete(d *schema.ResourceData, meta interface{}) error {
+	cluster_id := d.Get("cluster_id").(string)
+	topic := d.Get("topic").(string)
+	log.Printf("[INFO] Deleting Kafka topic %s in %s.", topic, cluster_id)
+	client := meta.(*Config).Client
+
+	err := client.DeleteKafkaTopic(cluster_id, topic)
+	if err != nil {
```

Comment:
If a Terraform created Kafka topic is deleted outside of Terraform, is deleting the topic through Terraform handled gracefully?

Reply:
Oh, I haven't tried that, need to have a quick try.